feat: Use probe-rs instead of probe-run

winksaville opened the pull request:
The probe-run carte has been deprecated,
https://ferrous-systems.com/blog/probe-run-deprecation. I followed the migration steps in the Blog Post
https://ferrous-systems.com/blog/probe-run-deprecation/#migration-steps

With this change all examples are functional, but if you have a version of probe-rs <= 0.22.0 then you may see an error; see: https://github.com/probe-rs/probe-rs/issues/2072. As a side note I used `screen /dev/ttyACM0` on my Arch Linux box to test `serial_loopback` and `serial_terminal`.

schteve commented:
Thanks for this! Could you please make one small change before I merge it? I'd like to bump the cortex-m from 0.7.5 to 0.7.7 so that the critical section feature works for the specified version. I think in practice it doesn't make a difference right now since the transitive dependencies resolve to 0.7.7 anyway, but it would be nice for it to be consistent.

Also thanks for raising and linking the probe-rs ticket about the error message. I was confused by that error while trying out probe-rs with this branch but you've already nipped that in the bud..

winksaville commented:
> Thanks for this! Could you please make one small change before I merge it? I'd like to bump the cortex-m from 0.7.5 to 0.7.7 so that the critical section feature works for the specified version. I think in practice it doesn't make a difference right now since the transitive dependencies resolve to 0.7.7 anyway, but it would be nice for it to be consistent.
> 

Do you want the change squashed and forced pushed or a separate commit?

> Also thanks for raising and linking the probe-rs ticket about the error message. I was confused by that error while trying out probe-rs with this branch but you've already nipped that in the bud..

NP!

schteve commented:
Either way is fine, I will probably squash it when I do the merge anyway.

winksaville commented:
Ok, I'll squash.

Another topic, I looked into "upgrading" dependencies, in particular embedded-hal and stm32F4xx-hal are quite out-of-date?
```
wink@3900x 24-01-26T03:50:58.883Z:~/prgs/rust/forks/nucleo-f446re (Use-probe-rs)
$ cargo upgrade
    Updating 'https://github.com/rust-lang/crates.io-index' index
    Checking nucleo-f446re's dependencies
name          old req compatible latest new req note        
====          ======= ========== ====== ======= ====        
cortex-m      0.7.5   0.7.7      0.7.7  0.7.7               
embedded-hal  0.2.7   0.2.7      1.0.0  0.2.7   incompatible
stm32f4xx-hal 0.13.2  0.13.2     0.20.0 0.13.2  incompatible
cortex-m-rt   0.7.1   0.7.3      0.7.3  0.7.3               
panic-probe   0.3.0   0.3.1      0.3.1  0.3.1               
   Upgrading recursive dependencies
note: Re-run with `--incompatible` to upgrade incompatible version requirements
note: Re-run with `--verbose` to show all dependencies
  unchanged: switch-hal, unwrap-infallible
```

As it turns out, there's some work to do, so I suggest "upgrading" be a separate change, if you want to do it at all. Personally, I like being up to date as possible, but you'll have to decide "what's possible" :)
```
wink@3900x 24-01-26T03:53:23.028Z:~/prgs/rust/forks/nucleo-f446re (Use-probe-rs)
$ cargo upgrade --incompatible
    Updating 'https://github.com/rust-lang/crates.io-index' index
    Checking nucleo-f446re's dependencies
name          old req compatible latest new req
====          ======= ========== ====== =======
embedded-hal  0.2.7   0.2.7      1.0.0  1.0.0  
stm32f4xx-hal 0.13.2  0.13.2     0.20.0 0.20.0 
   Upgrading recursive dependencies
note: Re-run with `--verbose` to show all dependencies
  unchanged: cortex-m, cortex-m-rt, panic-probe, switch-hal, unwrap-infallible
wink@3900x 24-01-26T04:01:04.070Z:~/prgs/rust/forks/nucleo-f446re (Use-probe-rs)
$ cargo run --example blinky
   Compiling syn v2.0.48
   Compiling stm32f4 v0.15.1
   Compiling embedded-hal v1.0.0
   Compiling litrs v0.4.1
   Compiling stm32f4xx-hal v0.20.0
   Compiling embedded-storage v0.3.1
   Compiling cortex-m-rt v0.7.3
   Compiling embedded-hal-nb v1.0.0
   Compiling document-features v0.2.8
   Compiling enumflags2_derive v0.7.8
   Compiling enumflags2 v0.7.8
   Compiling nucleo-f446re v0.1.0 (/home/wink/prgs/rust/forks/nucleo-f446re)
error[E0277]: the trait bound `stm32f4xx_hal::gpio::Pin<'A', 5, stm32f4xx_hal::gpio::Alternate<_>>: stm32f4xx_hal::timer::Pins<stm32f4xx_hal::pac::TIM2>` is not satisfied
   --> src/led.rs:69:38
    |
69  |         let mut pwm_ch1 = tim.pwm_hz(pin.into_alternate(), 20.kHz(), clocks).split();
    |                               ------ ^^^^^^^^^^^^^^^^^^^^ the trait `stm32f4xx_hal::timer::Pins<stm32f4xx_hal::pac::TIM2>` is not implemented for `stm32f4xx_hal::gpio::Pin<'A', 5, stm32f4xx_hal::gpio::Alternate<_>>`
    |                               |
    |                               required by a bound introduced by this call
    |
    = help: the following other types implement trait `stm32f4xx_hal::timer::Pins<TIM>`:
              ChannelBuilder<TIM, stm32f4xx_hal::::timer::pwm::{impl#31}::{constant#0}, NC1, O1>
              ChannelBuilder<TIM, stm32f4xx_hal::::timer::pwm::{impl#32}::{constant#0}, NC2, O2>
              ChannelBuilder<TIM, stm32f4xx_hal::::timer::pwm::{impl#33}::{constant#0}, NC3, O3>
              ChannelBuilder<TIM, stm32f4xx_hal::::timer::pwm::{impl#34}::{constant#0}, NC4, O4>
              (ChannelBuilder<TIM, stm32f4xx_hal::::timer::pwm::{impl#20}::{constant#0}, NC1, O1>, ChannelBuilder<TIM, stm32f4xx_hal::::timer::pwm::{impl#20}::{constant#2}, NC2, O2>, ChannelBuilder<TIM, stm32f4xx_hal::::timer::pwm::{impl#20}::{constant#4}, NC3, O3>, ChannelBuilder<TIM, stm32f4xx_hal::::timer::pwm::{impl#20}::{constant#6}, NC4, O4>)
              (ChannelBuilder<TIM, stm32f4xx_hal::::timer::pwm::{impl#21}::{constant#0}, NC2, O2>, ChannelBuilder<TIM, stm32f4xx_hal::::timer::pwm::{impl#21}::{constant#2}, NC3, O3>, ChannelBuilder<TIM, stm32f4xx_hal::::timer::pwm::{impl#21}::{constant#4}, NC4, O4>)
              (ChannelBuilder<TIM, stm32f4xx_hal::::timer::pwm::{impl#22}::{constant#0}, NC1, O1>, ChannelBuilder<TIM, stm32f4xx_hal::::timer::pwm::{impl#22}::{constant#2}, NC3, O3>, ChannelBuilder<TIM, stm32f4xx_hal::::timer::pwm::{impl#22}::{constant#4}, NC4, O4>)
              (ChannelBuilder<TIM, stm32f4xx_hal::::timer::pwm::{impl#23}::{constant#0}, NC1, O1>, ChannelBuilder<TIM, stm32f4xx_hal::::timer::pwm::{impl#23}::{constant#2}, NC2, O2>, ChannelBuilder<TIM, stm32f4xx_hal::::timer::pwm::{impl#23}::{constant#4}, NC4, O4>)
            and 7 others
note: required by a bound in `pwm_hz`
   --> /home/wink/.cargo/registry/src/index.crates.io-6f17d22bba15001f/stm32f4xx-hal-0.20.0/src/timer/pwm.rs:214:15
    |
212 |     fn pwm_hz<PINS>(self, pins: PINS, freq: Hertz, clocks: &Clocks) -> PwmHz<Self, PINS>
    |        ------ required by a bound in this associated function
213 |     where
214 |         PINS: Pins<Self>;
    |               ^^^^^^^^^^ required by this bound in `_stm32f4xx_hal_timer_PwmExt::pwm_hz`

error[E0599]: the method `split` exists for struct `PwmHz<TIM2, Pin<'A', 5, Alternate<_>>>`, but its trait bounds were not satisfied
   --> src/led.rs:69:78
    |
69  |         let mut pwm_ch1 = tim.pwm_hz(pin.into_alternate(), 20.kHz(), clocks).split();
    |                                                                              ^^^^^ method cannot be called on `PwmHz<TIM2, Pin<'A', 5, Alternate<_>>>` due to unsatisfied trait bounds
    |
   ::: /home/wink/.cargo/registry/src/index.crates.io-6f17d22bba15001f/stm32f4xx-hal-0.20.0/src/gpio.rs:243:1
    |
243 | pub struct Pin<const P: char, const N: u8, MODE = DefaultMode> {
    | --------------------------------------------------------------
    | |
    | doesn't satisfy `_: Pins<TIM2>`
    | doesn't satisfy `_: Split`
    |
    = note: the following trait bounds were not satisfied:
            `stm32f4xx_hal::gpio::Pin<'A', 5, stm32f4xx_hal::gpio::Alternate<_>>: stm32f4xx_hal::timer::Pins<stm32f4xx_hal::pac::TIM2>`
            `stm32f4xx_hal::gpio::Pin<'A', 5, stm32f4xx_hal::gpio::Alternate<_>>: stm32f4xx_hal::timer::pwm::sealed::Split`
            which is required by `stm32f4xx_hal::gpio::Pin<'A', 5, stm32f4xx_hal::gpio::Alternate<_>>: stm32f4xx_hal::timer::Pins<stm32f4xx_hal::pac::TIM2>`
            `stm32f4xx_hal::gpio::Pin<'A', 5, stm32f4xx_hal::gpio::Alternate<_>>: stm32f4xx_hal::timer::pwm::sealed::Split`

Some errors have detailed explanations: E0277, E0599.
For more information about an error, try `rustc --explain E0277`.
error: could not compile `nucleo-f446re` (lib) due to 2 previous errors
warning: build failed, waiting for other jobs to finish...
```

schteve commented:
This project probably isn't as useful with old dependencies that won't work with the latest common crates, so I'm in favor of updating them. In particular I think the embedded HAL interface going to 1.0 is significant and deserves an update. Definitely a separate PR though. Feel free to take a look if you like, but I might have some time soon.

Thanks for your help!

winksaville commented:
> This project probably isn't as useful with old dependencies that won't work with the latest common crates, so I'm in favor of updating them. In particular I think the embedded HAL interface going to 1.0 is significant and deserves an update. Definitely a separate PR though. Feel free to take a look if you like, but I might have some time soon.
> 
> Thanks for your help!

You're welcome thanks for doing this!
